### PR TITLE
Support weight-based eviction

### PR DESCRIPTION
**Pull request by tatsuya6502**
- Add support for weight based (relative size based) eviction.
    - Implement "Aggregate Victims" (AV) algorithm introduced by this paper: [Lightweight Robust Size Aware Cache Management](https://arxiv.org/abs/2105.08770) by Gil Einziger, Ohad Eytan, Roy Friedman, Ben Manes
- Add support for unbound cache.

## Edit on Dec 31, 2021

Finished implementing above features on the all caches:

- `future::Cache`
- `sync::Cache`
- `sync::SegmentedCache`
- `unsync::Cache`

#### API:

- `Cache` and `SegmentedCache`
    - Changed the return type of `max_capacity()` from `usize` to `Option<usize>`.
    - Added `builder()` method.
- `CacheBuilder`
    - Made the max capacity optional.
    - Added `weigher` method to set the weigher closure.
        
#### Internal Changes (Highlighting major changes):

- `sync::BaseCache`
    - Updated `do_insert_with_hash` method to calculate the policy weight of the entry using the weigher closure.
- `sync::InnerCache`
    - Changed to track the total weighed size.
    - Completely rewritten `admit` and `find_cache_victim` methods to do the weight-based admission.
    - Added `evict_lru_entries` method to evict LRU entries when the weighed size exceeds the max capacity.
        - This is needed to handle a case when an existing admitted entry was updated with a larger policy weight.
- `sync::ValueEntry`
    - Updated to hold the policy weight of the entry via `EntryInfo` enum.
    - Added `EntryInfo` enum to reduce the number of `Arc` usages:
        - This may have positive impacts on memory consumption and performance.
        - But also this may have negative impacts on performance as it does dynamic dispatches.

### Notes

- We still need to evaluate the aggregated victim strategy to see if it will improve cache hit rate in various real-world workloads (e.g. by using Caffeine's simulator). See [this comment](https://github.com/moka-rs/moka/pull/24#commitcomment-54926612) for details.

### CODE REVIEW

File: src/common.rs [hunk 1]
```diff
@@ -6,6 +6,8 @@ pub(crate) mod frequency_sketch;
 pub(crate) mod thread_pool;
 pub(crate) mod unsafe_weak_pointer;
 
+pub(crate) type Weighter<K, V> = Box<dyn Fn(&K, &V) -> u64 + Send + Sync + 'static>;
```
Comment (ben-manes) [inline src/common.rs:9]: In caffeine this is `Weigher`. I think this is a typo?

Comment (tatsuya6502) [reply src/common.rs:9]: Yes, it is a typo. Thank you for letting me know.

**Comment by tatsuya6502** [comment]
Finally, I have got some time to work on this PR. Now it is ready to merge.

**Review by tatsuya6502 — COMMENTED** [review]
Merging.

Although all unit tests are passing and I have reviewed the diff several times, I am a bit worried for regressions because this PR is huge (+2,009 -539). I will not release this immediately, but will spend next few days to run additional tests to shake out some bugs.

**Pull request merged**